File: epi_judge_python/list_cyclic_right_shift.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def cyclically_right_shift_list(L: ListNode, k: int) -> Optional[ListNode]:
    if not L:
        return None
    it2 = it1 = L
    length = 0
    while it1:
        length += 1
        it1 = it1.next 
    it1 = L
    for _ in range(k % length):
        it2 = it2.next
    if it1 is it2:
        return L
    while it2.next:
        it1 = it1.next
        it2 = it2.next
    st = it1.next
    it1.next = None
    it2.next = L
    return st

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('list_cyclic_right_shift.py',
                                       'list_cyclic_right_shift.tsv',
                                       cyclically_right_shift_list))

File: epi_judge_cpp/list_cyclic_right_shift.cc
#include <memory>

#include "list_node.h"
#include "test_framework/generic_test.h"
using std::shared_ptr;

shared_ptr<ListNode<int>> CyclicallyRightShiftList(shared_ptr<ListNode<int>> L,
                                                   int k) {
  auto dummy = std::make_shared<ListNode<int>>(0, L);
  auto shifter{dummy};
  if (L && k) {
    int count{0};
    while (shifter->next) {
      shifter = shifter->next;
      count++;
    }
    auto rotation = count - (k % count);
    if (rotation < count) {
      shifter = dummy;
      while (rotation--) {
        shifter = shifter->next;
      }
      auto temp{L};
      L = shifter->next;
      shifter->next.reset();
      shifter = L;
      while (shifter->next) {
        shifter = shifter->next;
      }
      shifter->next = temp;
    }
  }
  return L;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L", "k"};
  return GenericTestMain(
      args, "list_cyclic_right_shift.cc", "list_cyclic_right_shift.tsv",
      &CyclicallyRightShiftList, DefaultComparator{}, param_names);
}
